projects/stone_cx/robot.cc

```diff
@@ -1,170 +1,413 @@
+    }
+    
+    virtual void drive(float left, float right, bool log) override
+    {
+        const float length = (left * left) + (right * right);
+
+        // Steer based on signal
+        const float steering = (left - right) / length;
+        if(log) {
+            LOGI << "Steer:" << steering;
+        }
+
+        // Clamp absolute steering value to [0,1] and subtract from 1
+        // So no forward  speed if we're turning fast
+        const float forward = 1.0f - std::min(1.0f, std::fabs(steering));
+
+        m_Omni.omni2D(forward * 0.5f, 0.0f, steering * 8.0f);
+    }
+    
+private:
+    Robots::Omni2D &m_Omni;
+};
+
+std::unique_ptr<MotorController> createMotorController(Robots::Tank &tank)
+{
+    return std::make_unique<MotorControllerTank>(tank);
+}
+
+std::unique_ptr<MotorController> createMotorController(Robots::Omni2D &omni)
+{
+    return std::make_unique<MotorControllerOmni2D>(omni);
 }
-}   // Anonymous namespace
 
+void readHeadingThreadFunc(HeadingSource *headingSource,
+                           std::atomic<bool> &shouldQuit,
+                           std::atomic<float> &heading,
+                           unsigned int &numSamples)
+{
+    // While quit signal isn't set
+    for (numSamples = 0; !shouldQuit; numSamples++) {
+        heading = headingSource->getHeading().value();
+    }
+}
+}   // Anonymous namespace
 
-int main(int argc, char *argv[])
+int bob_main(int argc, char *argv[])
 {
-    const float velocityScale = 1.0f / 30.0f;
+    // Simulation rendering parameters
+    constexpr unsigned int activityImageWidth = 500;
+    constexpr unsigned int activityImageHeight = 1000;
+    const bool doVisualise = (argc > 1) && strcmp(argv[1], "--visualise") == 0;
+
+    // Create motor interface
+    Robots::ROBOT_TYPE motor;
+
+    std::unique_ptr<SpeedSource> speedSource = std::make_unique<SpeedSourceOmni2DDeadReckon>(motor);
+    std::unique_ptr<MotorController> motorController = createMotorController(motor);
+    
+#ifdef USE_EV3
+    std::unique_ptr<HeadingSource> headingSource = std::make_unique<HeadingSourceEV3IMU>(motor.getConnection());
+#else
+    std::unique_ptr<Net::Server> server;
+    std::unique_ptr<Net::Connection> connection;
+    std::unique_ptr<Video::NetSink> netSink;
+    std::unique_ptr<HeadingSource> headingSource = std::make_unique<HeadingSourceBN055IMU>();
+
+    // If command line arguments are specified, run connection
+    if(doVisualise) {
+        LOGI << "Streaming activity over network";
+
+        // Create server and sink for sending activity image over network
+        server = std::make_unique<Net::Server>();
+        connection = std::make_unique<Net::Connection>(server->waitForConnection());
+        netSink = std::make_unique<Video::NetSink>(*connection,
+                                                   cv::Size(activityImageWidth, activityImageHeight),
+                                                   "activity");
+        connection->runInBackground();
+    }
+#endif
```

Comment:
There should be a macro ``ROBOT_TYPE_EV3`` defined when using the EV3 robot, so we shouldn't need ``USE_EV3`` anywhere now.

Reply:
...and while we're on the subject of ``ROBOT_TYPE``s, FYI there is a ``BundledTankNetSink`` class which sets up the ``Server`` and listens etc. without any extra logic needed. It's a bit of a hack, but it does make it easy to switch between remote and non-remote robots. (Not saying you should necessarily use it here though)